Self propelled vehicles

Y-Pazniak opened the pull request:
The branch contains self-propelled vehicles and other fixes after replacement old code and add new functions of filling and loading tree from java-code

Y-Pazniak committed:
Added two empty branches.
M1 branch works with new branches - need to fill another branches.

Y-Pazniak committed:
Added two empty branches.
M1 branch works with new branches - need to fill another branches.

Y-Pazniak committed:
new step for "except_m1" branch has been added

Y-Pazniak committed:
weight selection for "except_m1 -> n1_n3" has been added

Y-Pazniak committed:
some prices for exceptM1 - n1-n3 has been added

Y-Pazniak committed:
Incorrect command ignoring was fixed.

Except m1 -> n1-n3 -> all branch with prices was added

Y-Pazniak committed:
Branch with prices for "except m1 -> m2-m3 -> electric engine" has been added

Y-Pazniak committed:
Branch with prices for "except m1 -> m2-m3 -> electric engine" has been added

Y-Pazniak committed:
Experiment with common engine type has failed.
Command tree json has been shorted using one common command for N1-N3 weights

Y-Pazniak committed:
Experiment with common engine type has failed.
Command tree json has been shorted using one common command for N1-N3 weights

Y-Pazniak committed:
branch "except m1 -> m2 -> engine volume" has been added WITHOUT price calculation

Y-Pazniak committed:
branch "except m1 -> m2 -> engine volume" has been finished

branch "except m1 -> truck units" has been started, types of trucks units has been added

Y-Pazniak committed:
branch "except m1 -> truck units" has been finished

Y-Pazniak committed:
branch "except m1" has been completely finished

Y-Pazniak committed:
Comments to the tree.json have been added.
More proper naming for nodes has been added.

Y-Pazniak committed:
Tree init has been removed from Bot.java
I've discovered we don't need it during bot loading - it loads with first "/start" command of any user and then we can use this singleton object anytime

Y-Pazniak committed:
Families for commands have been added

Y-Pazniak committed:
wrong path to properties was fixed

Y-Pazniak committed:
Adding steps into Command in process

Y-Pazniak committed:
Adding steps into Command has been finished

Y-Pazniak committed:
getUserPath array was replaced by list, no empty cell more

Y-Pazniak committed:
array has been replaced by list, rewriting of list has been added

Y-Pazniak committed:
passenger m1 cars useless type of owners were removed

Y-Pazniak committed:
bounding tree.json and step enum is in progress

Y-Pazniak committed:
binding tree.json and step enum for buses and trucks (passenger cars) 50% done

Y-Pazniak committed:
binding tree.json and step enum for buses and trucks (passenger cars) 50% done

Y-Pazniak committed:
binding tree.json and step enum for buses and trucks (passenger cars) has been finished

Y-Pazniak committed:
binding tree.json and step enum for buses and trucks (passenger cars) has been finished

Y-Pazniak committed:
Rewriting in list has been added

Y-Pazniak committed:
Rewriting in list has been added

Y-Pazniak committed:
Rewriting in list has been added

Y-Pazniak committed:
Rewriting in list has been added

Y-Pazniak committed:
Rewriting in list has been added

Y-Pazniak committed:
Rewriting in list has been added

Y-Pazniak committed:
Steps cleaners were deleted

Y-Pazniak committed:
bug with wrong sequence string was fixed

fixes from PR were added

Y-Pazniak committed:
tree has been updated - new branch (self-propelled vehicles) was added

Y-Pazniak committed:
adding self-propelled vehicles - choosing particular type is in progress

Y-Pazniak committed:
adding self-propelled vehicles - choosing particular type is in progress

Y-Pazniak committed:
second step for self-propelled vehicles has been added

Y-Pazniak committed:
help request has been added

Y-Pazniak committed:
price have been added

Y-Pazniak committed:
small visual improvements have been added

Y-Pazniak committed:
text formatting and bold for price have been added

Y-Pazniak committed:
mistake with wrong loop braking was fixed

Y-Pazniak committed:
useless strings from properties file were removed

Y-Pazniak committed:
useless tree has been removed

Y-Pazniak committed:
Constructor instead of creating object and filling setters has been added

Y-Pazniak committed:
fieldsToCommands instead of commandsToFields has been added.

Y-Pazniak committed:
commandsToFields Map mistake has been fixed

Y-Pazniak committed:
restored project from a great f*ck up

Y-Pazniak committed:
Tried to add a code of writing data into tree.json - for the first look it works fine

Y-Pazniak committed:
Started filling tree

Y-Pazniak committed:
Fixed mistakes, improved code, created a basement for next correct nodes filling in future

Y-Pazniak committed:
First branch calculation according to tree has been added. Needs testing!

Y-Pazniak committed:
Errors and useless code have been removed

Y-Pazniak committed:
Trucks and buses have been added, work fine.
Truck units have been added, but dont work - there is an error during weight setting, needs to be fixed

Y-Pazniak committed:
Trucks and buses have been added, work fine.
Truck units have been added, but dont work - there is an error during weight setting, needs to be fixed

Y-Pazniak committed:
All branch has been added succesfully

Y-Pazniak committed:
Families from command have been particulary removed

Y-Pazniak committed:
Families from command have been particulary removed

Y-Pazniak committed:
Families from command have been particulary removed

Y-Pazniak committed:
Families from command have been particulary removed

Y-Pazniak committed:
Families from command have been particulary removed

Y-Pazniak committed:
Families from command have been particulary removed

Y-Pazniak committed:
Families from command have been particulary removed

Y-Pazniak committed:
Families from command have been particulary removed

Y-Pazniak committed:
Families from command have been particulary removed

Y-Pazniak committed:
Graders have been added

Y-Pazniak committed:
Bulldozers have been added

Y-Pazniak committed:
project has been update for work with windows

Y-Pazniak committed:
filled some nods with a tool
outdated code has was removed

Y-Pazniak committed:
some new branches added

Y-Pazniak committed:
all nodes were filled

Y-Pazniak committed:
Use input stream stead of file

Y-Pazniak committed:
Stupid system variable naming has been removed

Y-Pazniak committed:
Unnecessary class has been removed

Y-Pazniak committed:
Unnecessary class has been removed

Y-Pazniak committed:
correct tree file name has been added

Y-Pazniak committed:
first try to get messages from tree

Y-Pazniak committed:
price from placeholders has been added

Y-Pazniak committed:
m1 (passenger cars) have been filled

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/UserProgress.java:
```diff
@@ -5,155 +5,156 @@
+
+        for (int i = 0; i < userPath.size(); i++) {
+            Command localCommand = userPath.get(i);
+            for (Map.Entry<StepsIndicator, Command> entry : fieldsToCommands.entrySet()) {
```
As we discussed, passing through a command tree is better than a map.

Y-Pazniak replied:
I did this code better, but I didn't manage to use the Tree here. The main reason - i didn't find the way how to use the tree to move back for several steps and to clean List<Command> userPath. 
Example. User path is: A.class -> B.class -> D.class -> E.class 
User wants to get back to the B.class. 
I need to fall back to B.class and clean all steps after it. I need to start a foreach and find an index of B.class object and then to clean this List userPath.sublist(b.index, userPath.size()).clean(); 
We can not use only the Tree for this purposes - we need foreach. 
Probably I did something wrong but I still don't know how to use the Tree for falling back user's path. We can discuss it.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/tree/ModifierTree.java:
```diff
@@ -0,0 +1,64 @@
+import java.io.IOException;
+import java.io.InputStream;
+
+public class ModifierTree {
```
It's more correct
```suggestion
public class TreeModifier {
```

Y-Pazniak replied:
Renamed. Should we rename also CommandTree to TreeCommand?

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/tree/ModifierTree.java:
```diff
@@ -0,0 +1,64 @@
+    public static String getPrice(final UserProgress userProgress) throws UtilsborCommandTreeReadingException {
+        Node root;
+        try {
+            root = mapper.readValue(getTreeInputStream(), Node.class);
```
I think it's redundant to build a tree here. I think it's enough to build a tree once on application starts up.

Y-Pazniak replied:
Done.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/tree/ModifierTree.java:
```diff
@@ -0,0 +1,64 @@
+                    if (node.getKey().equals(command)) {
+                        localNode = node;
+                        if (node.getPrice() != null) {
+                            return node.getPrice();
```
looks like you may return price even to not pass all user path.

Y-Pazniak replied:
Yes, my gross mistake. Fixed it

Y-Pazniak committed:
Fixed errors with incorrect price message sending.
Html markup removed from the tree.
The tree was prepared hor changing from String to Double

Y-Pazniak committed:
String price replaced to Double price

Y-Pazniak committed:
double price replacd to Price price (Price.class)

Y-Pazniak committed:
wrong method command validation naming changed (validateCommand -> isValidCommand).

Y-Pazniak committed:
Exception is expanded for a bundle request

Y-Pazniak committed:
Map fieldToCommand was replaced by code from tree and enums.

Y-Pazniak committed:
Map fieldToCommand was replaced by code from tree and enums.

Y-Pazniak committed:
TreeModifier was renamed to the proper name.
Redundant tree loading was removed.

Y-Pazniak committed:
Problem with returning price not null was fixed

Y-Pazniak committed:
UserPath changing replaced by "no-map code"

Y-Pazniak committed:
getCommandByStepIndicator has been added

Y-Pazniak committed:
getStepIndicatorByCommand has been added

Y-Pazniak committed:
getName for steps has been added

Y-Pazniak committed:
getName for steps has been added

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/tree/TreeModifier.java:
```diff
@@ -0,0 +1,56 @@
+                }
+            }
+        }
+        return "error during calculation";
```
For me, it looks like the exception better fits here.

Y-Pazniak committed:
double replaced by BigDecimal
useless map initialization removed

Y-Pazniak committed:
idk whats happening

Y-Pazniak committed:
ModifierTree was renamed to TreeModifier
